[GStreamer][MSE] Version check for a GStreamer bug fixed in 1.20.6

philn opened the pull request:
#### d54e09dcce0da58160346abf6ab95637313c1150
<pre>
[GStreamer][MSE] Version check for a GStreamer bug fixed in 1.20.6
<a href="https://bugs.webkit.org/show_bug.cgi?id=253169">https://bugs.webkit.org/show_bug.cgi?id=253169</a>

Reviewed by Alicia Boya Garcia.

Get the element state before activating a new source pad only in GStreamer versions older than
1.20.6. Newer versions no longer have this bug.

* Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
(webKitMediaSrcEmitStreams):

Canonical link: <a href="https://commits.webkit.org/261629@main">https://commits.webkit.org/261629@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c6d52c5866b210778d682b7e28af2652a8d83d4c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110304 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19397 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42941 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1708 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119265 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114253 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20853 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10578 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/1204 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116048 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15526 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98730 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102557 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97487 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30394 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43740 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12076 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31731 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85592 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state-changed (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12676 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8693 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18042 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51348 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8110 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14504 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
```diff
@@ -319,17 +319,17 @@ void webKitMediaSrcEmitStreams(WebKitMediaSrc* source, const Vector<RefPtr<Media
-
+        if (!webkitGstCheckVersion(1, 20, 6)) {
+            // Workaround: gst_element_add_pad() should already call gst_pad_set_active() if the element is PAUSED or
+            // PLAYING. Unfortunately, as of GStreamer 1.18.2 it does so with the element lock taken, causing a deadlock
```
> as of GStreamer 1.18.2

Now that the fix is released the comment should be updated to "versions of GStreamer prior to 1.20.6 (stable) / 1.21.x (unstable)".

philn replied:
> This check would disable the workaround in versions of 1.21 prior to the fix, which are still affected by the bug.

1.21 is a development version, distros are not shipping development versions, only stable releases.

ntrrgc replied:
People still install development versions from time to time though.

philn replied:
Well. At some point we need to find a balance between keeping technical debt at bay and pleasing every wild setup across the globe...

> People still install development versions from time to time though.

Of an old development branch? I don't think we should support that.
We usually do development version checks for the current development branch, not for versions that have been shipped as stable.

ntrrgc replied:
1.20-1.21 is not that old. And in this case the consequence of a false positive is a deterministic deadlock.

ntrrgc replied:
Here you have an algorithm for checking versions taking into consideration unstable versions. I think this should be used in webKitGstCheckVersion().

```cpp
constexpr bool gnomeStyleVersionCheck(const int wantedMajor, const int wantedMinor, const int wantedMicro,
                                      const int systemMajor, const int systemMinor, const int systemMicro)
{
    if (systemMajor < wantedMajor)
        return false;
    if (systemMajor > wantedMajor)
        return true;
    assert(systemMajor == wantedMajor);

    if (systemMinor < wantedMinor)
        return false;
    if (wantedMinor % 2) {
        // Checking against an unstable wanted: we can use regular ordering.
        if (systemMinor > wantedMinor)
            return true;
    } else {
        // Checking against an stable wanted, e.g: 1.20.5.

        // Accept the next stable, e.g: 1.22.0
        if (systemMinor > wantedMinor + 1)
            return true;
        // Reject the unstable with the same base, e.g. 1.21.7.
        if (systemMinor == wantedMinor + 1)
            return false;
    }

    assert(systemMinor == wantedMinor);
    return systemMicro >= wantedMicro;
}

// Tests for checking against a stable version
static_assert(!gnomeStyleVersionCheck(1, 20, 5, 1, 19, 7));
static_assert(gnomeStyleVersionCheck(1, 20, 5, 1, 22, 0));
static_assert(gnomeStyleVersionCheck(1, 20, 5, 1, 22, 0));
static_assert(!gnomeStyleVersionCheck(1, 20, 5, 1, 21, 0));
static_assert(!gnomeStyleVersionCheck(1, 20, 5, 1, 21, 5));
static_assert(!gnomeStyleVersionCheck(1, 20, 5, 1, 21, 7));

// Tests for checking against an unstable version
static_assert(!gnomeStyleVersionCheck(1, 21, 2, 1, 21, 1));
static_assert(!gnomeStyleVersionCheck(1, 21, 2, 1, 20, 8));
static_assert(gnomeStyleVersionCheck(1, 21, 2, 1, 21, 2));
static_assert(gnomeStyleVersionCheck(1, 21, 2, 1, 23, 1));
static_assert(gnomeStyleVersionCheck(1, 21, 2, 1, 21, 3));
static_assert(gnomeStyleVersionCheck(1, 21, 2, 1, 22, 0));
```

philn replied:
What's wrong with the current implementation?

ntrrgc replied:
The current implementation will e.g. accept 1.21.0 for a version check of 1.20.5, even tho 1.21.0 may very well have been released before 1.20.5 and lack the relevant fixes.

philn replied:
Sorry but I don't see the need to over-engineer this function. People installing unstable old GStreamer versions should not expect stable runtime behavior...

thiblahute replied:
Fwiw, I think we should not try to support unreleased versions fmpov.

ntrrgc replied:
>  Fwiw, I think we should not try to support unreleased versions fmpov.

Yet we have version checks for unstable versions in several parts of the code :confused: 

```
$ ggrep 'webkitGstCheckVersion(1, [0-9]*[13579],'
WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp
bool GStreamerMediaEndpoint::initializePipeline()
    if (webkitGstCheckVersion(1, 21, 0)) {
UniqueRef<GStreamerDataChannelHandler> GStreamerMediaEndpoint::findOrCreateIncomingChannelHandler(GRefPtr<GstWebRTCDataChannel>&& dataChannel)
    if (!webkitGstCheckVersion(1, 21, 0))

WebCore/Modules/mediastream/gstreamer/GStreamerStatsCollector.cpp
static gboolean fillReportCallback(GQuark, const GValue* value, gpointer userData)
        if (webkitGstCheckVersion(1, 21, 0)) {
        if (webkitGstCheckVersion(1, 21, 0)) {

WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp
void registerWebKitGStreamerElements()
        if (webkitGstCheckVersion(1, 21, 0)) {
```

philn replied:
Looks like your checkout is outdated, the only file that needs updating is Source/WebCore/Modules/mediastream/gstreamer/GStreamerStatsCollector.cpp

philn replied:
> Fwiw, I think we should not try to support unreleased versions fmpov.

I don't think that's avoidable when we want to use functionality not shipped in a stable gst release yet though.

ntrrgc replied:
I'll accept the patch...

ntrrgc reviewed (changes requested):
This check would disable the workaround in versions of 1.21 prior to the fix, which are still affected by the bug.

ntrrgc reviewed (approved):
r+

philn committed:
[GStreamer][MSE] Version check for a GStreamer bug fixed in 1.20.6
https://bugs.webkit.org/show_bug.cgi?id=253169

Reviewed by Alicia Boya Garcia.

Get the element state before activating a new source pad only in GStreamer versions older than
1.20.6. Newer versions no longer have this bug.

* Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
(webKitMediaSrcEmitStreams):

Canonical link: https://commits.webkit.org/261629@main
